docs/architecture/adr-102-rpc-companion.md

```diff
@@ -0,0 +1,358 @@
+# ADR-102: RPC Companion
+
+## Changelog
+
+- 2022-03-27: First draft (@andynog)
+
+## Status
+
+Accepted | Rejected | Deprecated | Superseded by
+
+## Context
+
+This ADR proposes an architecture of an ***RPC Companion*** solution, an instance of the proposed [ADR-101 Data Companion Pull API](https://github.com/cometbft/cometbft/pull/82).
+
+This solution can run as a sidecar which is a separate process that runs concurrently with the full node, but the RPC Companion is optional meaning that the full node will still provide RPC services that can be queried directly if operators don't want to run
+an RPC Companion service.
+
+This ADR provides a reference implementation of a system that can be used to offload queryable data from a CometBFT
+full node to a database and offer a service exposing the same JSONRPC methods on a endpoint as the regular JSONRPC methods of a CometBFT node endpoint,
+which makes it easier for integrators of RPC clients such as client libraries and applications to switch to this
+***RPC Companion*** with as minimum effort as possible.
+
+This architecture also makes it possible to scale horizontally the querying capacity of a full node by running multiple
+copies of the ***RPC Companion*** server instances that can be behind a scalable load-balancer (e.g. Cloudflare) which
```

Comment:
Would you happen to know how it's done right now? How do operators scale their API? Have we conducted any user interviews? Are there any issues from users related to the theme of this ADR?

Reply:
most operators might use a proxy server (Nginx) or a cloud load balancer (e.g. AWS ELB). Scaling the API is not just a matter of load balancing servers but eventually having a caching layer, geographical distribution, etc. The scalability topic is very complex and it is beyond the scope of this ADR in my opinion